HolyshiftService/RegistrationController.cs

```diff
@@ -0,0 +1,69 @@
+using Konscious.Security.Cryptography;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Newtonsoft.Json;
+using System.Security.Cryptography;
+using System.Text;
+
+public class RegistrationController : Controller
+{
+    private readonly HolyShiftDbContext _context;
+    private readonly JwtService _jwtService;
+
+    public RegistrationController(HolyShiftDbContext context, JwtService jwtService)
+    {
+        _context = context;
+        _jwtService = jwtService;
+    }
+
+    [HttpPost("api/register")]
+    public async Task<IActionResult> Register()
+    {
+        var requestBody = await new StreamReader(Request.Body).ReadToEndAsync();
+        var requestData = JsonConvert.DeserializeObject<Dictionary<string, string>>(requestBody);
+        if (requestData == null || 
+            !requestData.TryGetValue("email", out var email) || 
+            !requestData.TryGetValue("password", out var password) ||
+            !requestData.TryGetValue("userName", out var userName))
+        {
+            return BadRequest(new { Message = "Invalid request body" });
+        }
+        
+        var existingUser = await _context.Users.FirstOrDefaultAsync(u => u.Email == email);
+
+        if (existingUser != null)
+        {
+            return BadRequest(new { Message = "Email is already registered." });
+        }
+
+        byte[] salt = new byte[16];
+        using (var rng = RandomNumberGenerator.Create())
+        {
+            rng.GetBytes(salt);
+        }
+
+        using (var hasher = new Argon2id(Encoding.UTF8.GetBytes(password)))
+        {
+            hasher.Salt = salt;
+            hasher.DegreeOfParallelism = 8;
+            hasher.MemorySize = 65536;
+            hasher.Iterations = 4;
+
+            var hashedPassword = Convert.ToBase64String(hasher.GetBytes(32));
+
+            var user = new UserDbModel
+            {
+                Id = Guid.NewGuid(), 
+                UserName = userName, 
+                Email = email,
+                Salt = Convert.ToBase64String(salt),
+                PasswordHash = hashedPassword
+            };
+
+            _context.Users.Add(user);
+            await _context.SaveChangesAsync();
```

Comment:
In general there are multiple design patterns that abstract DB access and translate between DB layer and Business layer.

The most common 2 I encountered are `Data Access Object` and `Repository` patterns. Technically you can use both as they solve different problems.

For now let's go for DAO.

How it works is have a Dao interface/class per table (typically) with the typical access patterns the DB handles. And you use those access patterns.

e.g.
```C#
public interface IUserDao
{
  void AddUser(UserDbModel user);
  UserDbModel GetUserById(Guid id);
  // ...
}
```

Dao abstract the data access layer. So if we decide to change the access layer but the access patters are the same, we only need to inject a different interface. (e.g. changing from efcore to dapper)
Also Daos makes testing a lot simpler.